src/plugins/intel_npu/tests/unit/npuw/serialization.cpp

```diff
@@ -0,0 +1,272 @@
+
+    auto options_desc(std::make_shared<::intel_npu::OptionsDesc>());
+    options_desc->add<::intel_npu::NPUW_LLM_BATCH_DIM>();
+    ::intel_npu::Config var(options_desc);
+    ::intel_npu::Config res(options_desc);
+
+    std::map<std::string, std::string> tmp;
+    tmp["NPUW_LLM_BATCH_DIM"] = "42";
+    var.update(tmp);
+
+    std::stringstream ss;
+
+    write(ss, var);
+    read(ss, res);
+
+    EXPECT_EQ(res.get<::intel_npu::NPUW_LLM_BATCH_DIM>(), 42);
+}
+
+TEST(SerializationTest, BasicTypes_Any) {
+    using namespace ov::npuw::s11n;
+
+    int tmp = 42;
+    ov::Any var(tmp);
+    ov::Any res;
+
+    std::stringstream ss;
+
+    write_any(ss, var);
+    read_any(ss, res);
+
+    EXPECT_EQ(var.as<int>(), res.as<int>());
+}
+
+TEST(SerializationTest, BasicTypes_array) {
+    using namespace ov::npuw::s11n;
+
+    std::array<uint8_t, 6> res;
+
+    std::stringstream ss;
+
+    write(ss, NPUW_SERIALIZATION_INDICATOR);
+    read(ss, res);
+
+    EXPECT_EQ(NPUW_SERIALIZATION_INDICATOR, res);
+}
+
+TEST(SerializationTest, BasicTypes_pair) {
+    using namespace ov::npuw::s11n;
+
+    std::pair<int, float> var{42, 3.14};
+    std::pair<int, float> res;
+
+    std::stringstream ss;
+
+    write(ss, var);
+    read(ss, res);
+
+    EXPECT_EQ(var, res);
+}
+
+TEST(SerializationTest, BasicTypes_vector) {
+    using namespace ov::npuw::s11n;
+
+    std::vector<int> var{1, 2, 3, 45};
+    std::vector<int> res;
+
+    std::stringstream ss;
+
+    write(ss, var);
+    read(ss, res);
+
+    EXPECT_EQ(var, res);
+}
+
+TEST(SerializationTest, BasicTypes_map) {
+    using namespace ov::npuw::s11n;
+
+    std::map<int, std::string> var{{1, "a"}, {2, "b"}};
+    std::map<int, std::string> res;
+
+    std::stringstream ss;
+
+    write(ss, var);
+    read(ss, res);
+
+    EXPECT_EQ(var, res);
+}
+
+TEST(SerializationTest, BasicTypes_unordered_set) {
+    using namespace ov::npuw::s11n;
+
+    std::unordered_set<std::string> var{"a", "b", "c"};
+    std::unordered_set<std::string> res;
+
+    std::stringstream ss;
+
+    write(ss, var);
+    read(ss, res);
+
+    EXPECT_EQ(var, res);
+}
+
+TEST(SerializationTest, BasicTypes_optional) {
+    using namespace ov::npuw::s11n;
+
+    std::optional<int> var = 1;
+    std::optional<int> res;
+
+    std::optional<int> var2 = std::nullopt;
+    std::optional<int> res2;
+
+    std::stringstream ss;
+
+    write(ss, var);
+    write(ss, var2);
+    read(ss, res);
+    read(ss, res2);
+
+    EXPECT_EQ(var, res);
+    EXPECT_EQ(var2, res2);
+}
+
+TEST(SerializationTest, BasicTypes_Tensor_with_weights) {
+    using namespace ov::npuw::s11n;
+
+    std::vector<uint8_t> data {0, 1, 2, 3};
+    ov::Tensor var(ov::element::u8, ov::Shape({2, 2}), data.data());
+    std::vector<ov::Tensor> res;
+
+    std::stringstream ss;
+
+    std::unordered_map<const void*, std::size_t> const_offset;
+    Context ctx(false, const_offset);
+
+    write_weightless(ss, {var}, ctx);
+    read_weightless(ss, res, nullptr);
```

Comment:
test says "with_weights"

Reply:
Yeah, but there is no way to test proper weightless mode - we don't have a model - thus no weights. `write_weightsless` does support both cases - it tries to write as weightless, but if it's not possible - it will serialize full `Tensor`

Reply:
I suggest to put a comment about that here